EWMH Sticky windows

Airblader opened the pull request:
fixes #1455

Airblader commented:
I realize that a test is still missing, but I'm having some problems executing tests locally at the moment. At least one test always fails, saying that i3 couldn't be killed (always a different test).

Opening the PR already anyway to allow reviewing it upfront (and maybe also for someone to play around with it).

acrisci commented on include/con.h:
```diff
@@ -56,6 +56,18 @@ bool con_is_split(Con *con);
+ * Make this container sticky and update the EWMH hint.
+ *
+ */
+void con_set_sticky(Con *con, bool sticky, bool update_ewmh);
```
Nonobvious boolean params is an anti-pattern. I would prefer to add an enum for the switch (and also in other places in the code where this is the case).

Airblader replied:
Why is it non-obvious? We can turn it into an enum, but all we have here is an on/off state, for which a boolean seems suitable to me.

acrisci replied:
I just searched for "boolean parameters" and read these two articles:

http://blog.ometer.com/2011/01/20/boolean-parameters-are-wrong/

http://ariya.ofilabs.com/2011/08/hall-of-api-shame-boolean-trap.html

Airblader replied:
Well, the first one states that a boolean parameter is OK if the method name describes it, which "set_sticky" does. However, I've later on added the second argument, which I guess does make it a little confusing. I'll create an enum for it. 

However, consequently this would mean having to introduce an enum for whether or not the EWMH hint should be updated – or let the caller handle it, which is how I had it at first, but putting it into the function forces the caller to think about whether or not it's needed. I'm not really sure yet which is the better approach.

acrisci replied:
That I think is a matter of style. The alternative would be to keep track that the property changed and then wait to do the actual X11 update during a tree render. That is the preferred way to do things in other parts of the code.

Airblader replied:
We don't do that for EWMH hints at all, I think. It'd also not be that easy here since we can't decide whether the update is necessary or not.

acrisci replied:
I personally have no problem with how it is now, but some comments in the code indicated Michael had some issues with this sort of thing in previous versions, so he may have an opinion.

Airblader replied:
OK. We'll see what he says.

acrisci commented:
I tested it and it seems to work ok for me.

However, the testcases failing is a serious issue and you need to look into why that is happening.

Airblader commented:
@acrisci Yeah, I can see now that the failing tests are caused by the PR, not my local setup. I'm trying to figure out what is going on here since I don't experience any problems running the patch on my machine.

Airblader commented:
OK, I found the problem. Fixing.

acrisci commented on docs/userguide:
```diff
@@ -1818,6 +1818,27 @@ bindsym $mod+c move absolute position center
+Some windows, such as desktop notifications, should stay visible even if
+another workspace is opened after the notification appeared. i3 supports the
+sticky window concept from the EWMH specification and will automatically
+respect it. However, sometimes it makes sense to manually make some window
```
I would ger rid of the first two sentences of this paragraph and expand on the rest. I don't think the EWMH standard is appropriate to discuss in the userguide. You should write from the perspective of a user and not a developer in this document.

Airblader replied:
Yes, you're right. I'll work on this and also add a more meaningful example.

acrisci commented on include/data.h:
```diff
@@ -593,6 +593,8 @@ struct Con {
     TAILQ_HEAD(swallow_head, Match) swallow_head;
 
     fullscreen_mode_t fullscreen_mode;
+    bool sticky;
```
Could use some documentation here about the property.

acrisci commented:
I don't like how when I switch desktops, the sticky window is always the focused window.

Airblader commented:
True, it shouldn't mess with the focus.

Airblader commented:
@acrisci I've made a few changes now. I also didn't introduce an enum since I got rid of that function altogether and hence any ambiguity. The EWMH hint only needs to be updated in a single place anyway.

I've also dealt with the focus problem. Maybe we want to talk about whether this is the right approach. I should also point out that the sticky window won't have focus even if it had it on the source workspace, but I think this is what we should expect.

Multiple sticky windows doesn't work yet, I'm going to investigate that next. And then, of course, I'll also write tests.

Airblader commented:
What if a window is made sticky which currently is on a workspace that isn't visible? Should it be immediately moved to the current workspace?

Airblader commented:
I updated the core part where the containers are moved to handle multiple sticky containers correctly now. I also changed the focus strategy a little: if a container was focused on the target workspace, it will still be focused. If the workspace was empty, the sticky container will be focused.

acrisci commented:
> What if a window is made sticky which currently is on a workspace that isn't visible? Should it be immediately moved to the current workspace?

Yeah, it should really be always on top of other windows, but it's probably not important to take the metaphor that far.

Airblader commented:
I've added a few tests now. Any wishes for other tests?

I'm not entirely done yet, though, because I have to get windows to jump to the current workspace if stickiness is enabled for a window on a non-visible workspace.

acrisci commented on src/commands.c:
```diff
@@ -1537,6 +1537,38 @@ void cmd_fullscreen(I3_CMD, char *action, char *fullscreen_mode) {
+        ewmh_update_sticky(current->con->window->id, sticky);
+    }
+
+    cmd_output->needs_tree_render = false;
```
You don't need to set this to false.

acrisci commented on testcases/t/251-sticky.t:
```diff
@@ -0,0 +1,85 @@
+cmd 'sticky enable';
+$ws = fresh_workspace;
+
+is(@{get_ws($ws)->{floating_nodes}}, 0, 'tiling sticky container did not move');
```
I think you meant to check for tiling nodes.

Airblader replied:
Well, yes and no. I did mean to write floating_nodes, but I think it would actually make sense to assert against both.

Airblader commented:
Would it make sense to try and find a feasible shortcut that isn't used in the default config and see if we can work it in there? Maybe `$mod+Shift+s`?

Airblader commented:
@acrisci @stapelberg OK, I've pushed another commit that pops sticky windows to the front if they are made sticky while not being visible. I realize that I've added yet another boolean argument and I'm happy to create an enum for it instead, but I want some feedback on the overall solution first.

There's also still the open point of whether we want to configure a default binding for this.

stapelberg commented on src/handlers.c:
```diff
@@ -706,6 +707,15 @@ static void handle_client_message(xcb_client_message_event_t *event) {
                 con_set_urgency(con, false);
             else if (event->data.data32[0] == _NET_WM_STATE_TOGGLE)
                 con_set_urgency(con, !con->urgent);
+        } else if (event->data.data32[1] == A__NET_WM_STATE_STICKY) {
```
Please add a log message here, so that we can diagnose the case where an application sets `_NET_WM_STATE_STICKY` and users are confused about what happens.

Airblader replied:
I guess I never answered, but I've added it (some time ago).

stapelberg commented on src/output.c:
```diff
@@ -46,3 +46,50 @@ Output *get_output_from_string(Output *current_output, const char *output_str) {
+    Con *output;
+    TAILQ_FOREACH(output, &(croot->nodes_head), nodes) {
+        /* We cannot simply iterate over the floating windows and move them as
+         * this will modify the list concurrently. Instead, we first figure out
```
What’s the problem with modifying the list while iterating? In commit bf3cd41b5ddf1e757515ab5fbf811be56e5f69cc, `reorder_bindings_of_mode()` does that.

Airblader replied:
Yeah, but there you only reorder, but don't make the list smaller. I don't think the concept there applies here and simply iterating messes up the loop and causes crashes. But since we don't remove every element from the list we also can't loop over the first element until the list is empty either.

stapelberg replied:
Sorry, I don’t quite follow. Why does the following not work?

```c
for (output = TAILQ_FIRST(&(croot->nodes_head)); output != TAILQ_END(&(croot->nodes_head)); ) {
Con *current = output;
output = TAILQ_NEXT(output, nodes);
// maybe call TAILQ_REMOVE
}
```

Airblader replied:
(Note that the loop in question is the one over the windows, not the outputs)
Mh, yeah I guess that would work, at least I don't see a problem with it now. I'll change it.

stapelberg commented on docs/userguide:
```diff
@@ -1818,6 +1818,28 @@ bindsym $mod+c move absolute position center
 bindsym $mod+m move position mouse
 -------------------------------------------------------
 
+=== Sticky containers
```
s/Sticky containers/Sticky floating windows/? It’s more specific than just containers, right?

stapelberg commented on include/con.h:
```diff
@@ -212,10 +218,14 @@ void con_disable_fullscreen(Con *con);
  */
-void con_move_to_workspace(Con *con, Con *workspace, bool fix_coordinates, bool dont_warp);
+void con_move_to_workspace(Con *con, Con *workspace, bool fix_coordinates,
+                           bool dont_warp, bool ignore_focus);
```
I’m wondering if it might be cleaner to make the callers deal with focus handling instead of introducing `ignore_focus`. Can you investigate please?

Airblader replied:
Given the complexity of it (there's multiple steps of focus involved at different points of the workflow), I don't think we'll do us any favor moving this to the client (or whether it's possible at all).

stapelberg commented:
The overall approach seems fine to me.

Airblader commented:
@stapelberg I addressed all your comments.

stapelberg commented on docs/userguide:
```diff
@@ -1834,6 +1834,28 @@ bindsym $mod+c move absolute position center
+window.
+
+Note that while any window can be made sticky through this command, it will
+only take effect if the window is floating.
```
s/it/stickyness/ for clarity.

Should we add a reference to issue #1758 for stickyness in tiling containers?

Airblader replied:
I don't think the documentation should refer to the issue tracker at all.

stapelberg replied:
Why? It currently just silently omits the question that readers might have.

Airblader replied:
IMHO, the documentation should just describe the state as-is, not what the future may or may not hold. In the end I don't care that much, if you want me to add a reference, I'll do so.

stapelberg commented on docs/userguide:
```diff
@@ -1834,6 +1834,28 @@ bindsym $mod+c move absolute position center
+*Examples*:
+---------------------------------------------
+# make a terminal sticky that was started as
+# a notepad
```
Is the line-wrap necessary here? I think the 2 words would fit just fine in the previous line, don’t they?

Airblader replied:
True. Fixed.

stapelberg commented on testcases/t/251-sticky.t:
```diff
@@ -0,0 +1,100 @@
+# Ticket: #1455
+use i3test;
+
+sub kill_all {
```
I think this function is somewhat a misnomer. It doesn’t kill all windows, it just kills these with class=findme. Given that it’s only one command you’re wrapping, why not replace the callers with just the `cmd '[class="findme"] kill'`? I wouldn’t be concerned with duplication, but rather value clarity for the reader.

Airblader replied:
Yeah, I think this was just a relict from before I made it simpler. Fixed.

Airblader commented:
@stapelberg OK, I've changed the loop construct now. Looks a lot better, thanks. :9

stapelberg commented on src/output.c:
```diff
@@ -46,3 +46,34 @@ Output *get_output_from_string(Output *current_output, const char *output_str) {
+        Con *workspace, *visible_ws = NULL;
+        GREP_FIRST(visible_ws, output_get_content(output), workspace_is_visible(child));
+
+        /* We use this loop instead of TAILQ_FOREACH to avoid problems if the sticky window was the last window on that workspace as
```
Can you wrap this comment to fit in 80 chars? I know it’s not a hard limit, but we typically wrap comments rather early. vim’s gqq feature will do the right thing.

stapelberg commented on src/output.c:
```diff
@@ -46,3 +46,34 @@ Output *get_output_from_string(Output *current_output, const char *output_str) {
+
+        /* We use this loop instead of TAILQ_FOREACH to avoid problems if the sticky window was the last window on that workspace as
+         * moving it in this case will close the workspace. */
+        for (workspace = TAILQ_FIRST(&(output_get_content(output)->nodes_head)); workspace != TAILQ_END(&(output_get_content(output)->nodes_head));) {
```
Here and below, can you add newlines after the `;` to make the loop more readable please?

stapelberg commented:
Can you resolve the merge conflicts please?

Airblader commented:
@stapelberg Done. I also won't get tired bringing up the default keybinding question. :smile:

stapelberg commented:
Sorry, I forgot answering that question. I don’t think this feature warrants a default keybinding. After all, it’s restricted to floating windows which we discourage in general.
